### PR TITLE
Add gpg-agent

### PR DESCRIPTION
**Pull request by adrien-delhorme**
The Duplicity package (PR #1810) requires gpg-agent in order to run. I tried to make a GnuPG package.
Only tested on a 88f6281.

### CODE REVIEW

File: cross/libassuan/Makefile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+PKG_NAME = libassuan
+PKG_VERS = 2.2.1
+PKG_EXT = tar.bz2
+PKG_DIST_NAME = $(PKG_NAME)-$(PKG_VERS).$(PKG_EXT)
+PKG_DIST_SITE = ftp://ftp.gnupg.org/gcrypt/$(PKG_NAME)
+PKG_DIR = $(PKG_NAME)-$(PKG_VERS)
+
+DEPENDS = cross/libgpg-error
+
+HOMEPAGE = https://www.gnupg.org/related_software/libassuan/index.html
```
Comment (Dr-Bean) [inline cross/libassuan/Makefile]: Caught my eye. `https://www.gnupg.org/related_software/libassuan` looks better ;)

File: cross/libgpg-error/Makefile [hunk 2]
```diff
@@ -11,6 +11,13 @@ HOMEPAGE = http://www.gnupg.org/related_software/libgpg-error/
...
 LICENSE  = LGPL
 
+PRE_CONFIGURE_TARGET = preConfigureTarget
 GNU_CONFIGURE = 1
 
 include ../../mk/spksrc.cross-cc.mk
+
+.PHONY: preConfigureTarget
+preConfigureTarget:
+	cp $(WORK_DIR)/$(PKG_DIR)/src/syscfg/lock-obj-pub.arm-unknown-linux-gnueabi.h $(WORK_DIR)/$(PKG_DIR)/src/syscfg/lock-obj-pub.linux-gnueabi.h
```
Comment (Dr-Bean) [inline cross/libgpg-error/Makefile]: Not sure about these two lines (L21-22). Seems to me this won't work for all arches, will it? I'll probably take a better look at it later.

File: cross/libksba/Makefile [hunk 3]
```diff
@@ -0,0 +1,17 @@
+PKG_NAME = libksba
+PKG_VERS = 1.3.3
+PKG_EXT = tar.bz2
+PKG_DIST_NAME = $(PKG_NAME)-$(PKG_VERS).$(PKG_EXT)
+PKG_DIST_SITE = ftp://ftp.gnupg.org/gcrypt/$(PKG_NAME)
+PKG_DIR = $(PKG_NAME)-$(PKG_VERS)
+
+DEPENDS = cross/libgpg-error
+
+HOMEPAGE = https://www.gnupg.org/related_software/libksba/index.html
```
Comment (Dr-Bean) [inline cross/libksba/Makefile]: Maybe remove `/index.html` here as well.

File: cross/npth/Makefile [hunk 4]
```diff
@@ -0,0 +1,16 @@
+PKG_NAME = npth
+PKG_VERS = 1.2
+PKG_EXT = tar.bz2
+PKG_DIST_NAME = $(PKG_NAME)-$(PKG_VERS).$(PKG_EXT)
+PKG_DIST_SITE = ftp://ftp.gnupg.org/gcrypt/$(PKG_NAME)
+PKG_DIR = $(PKG_NAME)-$(PKG_VERS)
+
+DEPENDS = cross/gnutls
+
+HOMEPAGE =
```
Comment (Dr-Bean) [inline cross/npth/Makefile]: Please add some info here and for `COMMENT` as well.

File: spk/gnupg/Makefile [hunk 5]
```diff
@@ -0,0 +1,37 @@
+SPK_NAME = gnupg
+SPK_VERS = 2.1.6
+SPK_REV = 1
+SPK_ICON = src/gnupg.png
+DSM_UI_DIR = app
```
Comment (Dr-Bean) [inline spk/gnupg/Makefile]: I think this can go: the package isn't startable.

File: spk/gnupg/Makefile [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+DEPENDS = cross/$(SPK_NAME)
+
+MAINTAINER = SynoCommunity
+DESCRIPTION = GnuPG allows to encrypt and sign your data and communication, features a versatile key management system as well as access modules for all kinds of public key directories.
+RELOAD_UI = yes
```
Comment (Dr-Bean) [inline spk/gnupg/Makefile]: Probably unneeded.

File: spk/gnupg/Makefile [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+HOMEPAGE = http://www.gnupg.org/
+LICENSE  = GNU General Public License
+
+INSTALLER_SCRIPT = src/installer.sh
+SSS_SCRIPT       = src/dsm-control.sh
+
+INSTALL_PREFIX = /usr/local/$(SPK_NAME)
+
+POST_STRIP_TARGET = gnupg_extra_install
```
Comment (Dr-Bean) [inline spk/gnupg/Makefile]: As long as DSM_UI_DIR is removed, this can be removed...

File: spk/gnupg/Makefile [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+include ../../mk/spksrc.spk.mk
+
+.PHONY: gnupg_extra_install
```
Comment (Dr-Bean) [inline spk/gnupg/Makefile]: ...as can L29-L37

File: spk/gnupg/src/installer.sh [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    exit 0
+}
+
+postinst ()
+{
+    # Link
+    ln -s ${SYNOPKG_PKGDEST} ${INSTALL_DIR}
+
+    # Add symlink
+    ln -s ${INSTALL_DIR}/bin/gpg-agent /usr/syno/bin/gpg-agent
```
Comment (Dr-Bean) [inline spk/gnupg/src/installer.sh]: We can't put symlinks into system locations. `/usr/local/bin` is as far as we can go.

File: spk/gnupg/src/installer.sh [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+preuninst ()
+{
+    exit 0
+}
+
+postuninst ()
+{
+    # Remove link
+    rm -f ${INSTALL_DIR}
+    rm -f /usr/syno/bin/gpg-agent
```
Comment (Dr-Bean) [inline spk/gnupg/src/installer.sh]: ...which means this needs changing too.

**Comment by Dr-Bean** [comment]
Added some comments, take a look.
One other thing: can you arrange a larger logo? 256x256 would be great, but 72x72 isn't going to cut it.

File: cross/libgpg-error/Makefile (hunk 2 above)
Comment (adrien-delhorme) [inline cross/libgpg-error/Makefile]: This probably won't work for all arches. But I really don't know how to improve that.
I had an error when cross compiling gnupg : ``./gpg-error.h.in:320: error including `./syscfg/lock-obj-pub.linux-gnueabi.h': No such file or directory``
I found this partial solution on https://github.com/OpenELEC/OpenELEC.tv/blob/9df5e4f92a113c7d2fdc2cff8a1757f70103299f/packages/security/libgpg-error/package.mk

File: cross/npth/Makefile (hunk 4 above)
Comment (adrien-delhorme) [inline cross/npth/Makefile]: I can't find an homepage for nPth. Should I add http://gnupg.org, or http://lwn.net/Articles/496268/?

Comment (Dr-Bean) [inline cross/npth/Makefile]: The first link is fine, it's the 'home' of the project.

File: cross/libgpg-error/Makefile (hunk 2 above)
Comment (Dr-Bean) [inline cross/libgpg-error/Makefile]: If you get an actual error during cross-compiling which tells you what name is required, then it seems easy to me: Why not just compile all arches (`make all-supported && `make all-legacy`), and jot down the file names that are generated, and the file names that are required?

If the names are all built up the same way, there's got to be a way to catch that in a couple of lines of code, while being as future proof as possible.

Your linked code, as well as the original code it was based off (http://cgit.openembedded.org/openembedded-core/tree/meta/recipes-support/libgpg-error/libgpg-error_1.19.bb#n23) can probably serve as inspiration.

Comment (Dr-Bean) [inline cross/libgpg-error/Makefile]: @adrien-delhorme https://github.com/SynoCommunity/spksrc/pull/1680 has a better way of handling the libgpg-error changes. That PR will be merged in the next couple of days. The easiest for you to do is to remove all libgpg-error related changes in this PR.

**Comment by adrien-delhorme** [comment]
I updated and rebased the pull request.

File: spk/gnupg/src/installer.sh [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    exit 0
+}
+
+postinst ()
+{
+    # Link
+    ln -s ${SYNOPKG_PKGDEST} ${INSTALL_DIR}
+
+    # Add symlink
+    ln -s ${INSTALL_DIR}/bin/gpg-agent /usr/local/bin/gpg-agent
```
Comment (Diaoul) [inline spk/gnupg/src/installer.sh]: Add `mkdir -p /usr/local/bin`

**Comment by adrien-delhorme** [comment]
Done

File: cross/gnupg/Makefile [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+PKG_VERS = 2.1.6
+PKG_EXT = tar.bz2
+PKG_DIR = $(PKG_NAME)-$(PKG_VERS)
+PKG_DIST_NAME = $(PKG_DIR).$(PKG_EXT)
+PKG_DIST_SITE = ftp://ftp.gnupg.org/gcrypt/$(PKG_NAME)
+
+DEPENDS = cross/libgpg-error cross/libgcrypt cross/libassuan cross/libksba cross/npth cross/gnutls
+
+HOMEPAGE = http://www.gnupg.org/
+COMMENT  =  GnuPG allows to encrypt and sign your data and communication, features a versatile key management system as well as access modules for all kinds of public key directories.
```
Comment (Dr-Bean) [inline cross/gnupg/Makefile]: Minor: one space too many here :)

File: spk/gnupg/Makefile [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+HOMEPAGE = http://www.gnupg.org/
+LICENSE  = GNU General Public License
+
+INSTALLER_SCRIPT = src/installer.sh
+SSS_SCRIPT       = src/dsm-control.sh
+
+INSTALL_PREFIX = /usr/local/$(SPK_NAME)
+
+include ../../mk/spksrc.spk.mk
```
Comment (Dr-Bean) [inline spk/gnupg/Makefile]: Could you add a newline please?

**Comment by Dr-Bean** [comment]
Should be good to go if the last couple of comments are addressed (all minor) :)

**Comment by adrien-delhorme** [comment]
Fixed. I think its good now.

**Comment by Dr-Bean** [comment]
gnupg 2.1.6-1 has been published on the repository as a beta package. Nice work :)

**Pull request merged**